src/passes/OptimizeInstructions.cpp

```diff
@@ -458,6 +458,91 @@ struct OptimizeInstructions
           return c;
         }
       }
+      {
+        Const* c;
+        Binary* bin;
+        Expression *x, *y;
+        BinaryOp op;
+
+        // fneg(x) * fneg(y)   ==>   x * y
+        // fneg(x) / fneg(y)   ==>   x / y
+        if (matches(
+              curr,
+              binary(&bin, Mul, unary(Neg, any(&x)), unary(Neg, any(&y)))) ||
+            matches(
+              curr,
+              binary(&bin, DivS, unary(Neg, any(&x)), unary(Neg, any(&y))))) {
+          bin->left = x;
+          bin->right = y;
+          return bin;
+        }
+        // fneg(x * C)   ==>   x * -C
+        // fneg(x / C)   ==>   x / -C
+        // fneg(C / x)   ==>   -C / x
+        //
+        // TODO: using `fastMath` is not necessary, but due to fuzzer can't
+        // flexible handle sign of NaNs we use it here.
+        if (fastMath &&
+            (matches(curr, unary(Neg, binary(&bin, Mul, any(&x), fval(&c)))) ||
+             matches(curr, unary(Neg, binary(&bin, DivS, any(&x), fval(&c)))) ||
+             matches(curr,
+                     unary(Neg, binary(&bin, DivS, fval(&c), any(&x)))))) {
+          c->value = c->value.neg();
+          return bin;
+        }
+        // (fneg(x) * y) op C   ==>   (x * y) op -C
+        // (x * fneg(y)) op C   ==>   (x * y) op -C
+        // (fneg(x) / y) op C   ==>   (x / y) op -C
+        // (x / fneg(y)) op C   ==>   (x / y) op -C
+        //   where op = `*` `/` or `+`
+        if ((matches(curr,
+                     binary(&op,
+                            binary(&bin, Mul, unary(Neg, any(&x)), any(&y)),
+                            fval(&c))) ||
+             matches(curr,
+                     binary(&op,
+                            binary(&bin, Mul, any(&x), unary(Neg, any(&y))),
+                            fval(&c))) ||
+             matches(curr,
+                     binary(&op,
+                            binary(&bin, DivS, unary(Neg, any(&x)), any(&y)),
+                            fval(&c))) ||
+             matches(curr,
+                     binary(&op,
+                            binary(&bin, DivS, any(&x), unary(Neg, any(&y))),
+                            fval(&c)))) &&
+            (op == Abstract::getBinary(bin->type, Mul) ||
+             op == Abstract::getBinary(bin->type, DivS) ||
+             op == Abstract::getBinary(bin->type, Add)) &&
+            !x->is<Const>() && !y->is<Const>() && !c->value.isNaN()) {
```

Comment:
```suggestion
            !x->is<Const>() && !y->is<Const>()) {
```
A NaN is not a problem here. The final operation done is `op` both before and after the optimization, which will lead to NaN nondeterminism that allows any sign NaN (and the fuzzer will pick a positive one).

Reply:
Ok, I will try to remove this NaNs